ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java

```diff
@@ -584,173 +540,100 @@ private FunctionDescriptionList describeFunction(final String statementText,
     );
   }
 
-  private QueryDescription explainQuery(final Explain explain, final String statementText) {
-    final String queryId = explain.getQueryId();
-    if (queryId != null) {
-      final PersistentQueryMetadata metadata =
-          ksqlEngine.getPersistentQuery(new QueryId(queryId));
-      if (metadata == null) {
-        throw new KsqlException(
-            "Query with id:" + queryId + " does not exist, use SHOW QUERIES to view the full "
-                + "set of queries.");
-      }
-      return QueryDescription.forQueryMetadata(metadata);
-    }
-    final QueryMetadata queryMetadata = getStatementExecutionPlan(
-        explain.getStatement(),
-        statementText,
-        Collections.emptyMap()
-    );
-    if (queryMetadata == null) {
-      throw new KsqlException("The provided statement does not run a ksql query");
-    }
-    return QueryDescription.forQueryMetadata(queryMetadata);
+  private void validateExecutableStatements(
+      final List<? extends PreparedStatement<?>> statements,
+      final Map<String, Object> propertyOverrides
+  ) {
+    final List<PreparedStatement<?>> withSchemas = statements.stream()
+        .map(this::addInferredSchema)
+        .collect(Collectors.toList());
+
+    ksqlEngine.tryExecute(withSchemas, ksqlConfig, propertyOverrides);
   }
 
-  private QueryMetadata getStatementExecutionPlan(
-      final Statement statement, final String statementText, final Map<String, Object> properties) {
+  private CommandStatusEntity distributeStatement(
+      final PreparedStatement<?> statement,
+      final Map<String, Object> propertyOverrides
+  ) {
+    try {
+      final PreparedStatement<?> withSchema = addInferredSchema(statement);
 
-    final KsqlStatementTask ksqlStatementTask = ksqlStatementTasks.get(statement.getClass());
-    if (ksqlStatementTask == null) {
-      throw new KsqlException("Cannot FIND execution plan for this statement:" + statement);
-    }
-    return ksqlStatementTask.execute(statement, statementText, properties);
-  }
+      final QueuedCommandStatus queuedCommandStatus = replayableCommandQueue.enqueueCommand(
+          withSchema.getStatementText(),
+          withSchema.getStatement(),
+          ksqlConfig,
+          propertyOverrides);
 
-  private interface KsqlStatementTask {
-    QueryMetadata execute(Statement statement, String statementText,
-                          Map<String, Object> properties);
+      final CommandStatus commandStatus = queuedCommandStatus
+          .tryWaitForFinalStatus(Duration.ofMillis(distributedCommandResponseTimeout));
+
+      return new CommandStatusEntity(
+          withSchema.getStatementText(),
+          queuedCommandStatus.getCommandId(),
+          commandStatus
+      );
+    } catch (final Exception e) {
+      throw new KsqlException(String.format(
```

Comment:
Not part of your PR I know, but we shouldn't convert to a KsqlException here.

Reply:
`KsqlException` is more for user facing errors, right?  (We should document that in `KsqlExeption`s javadocs if thats the case.

What _should_ we be throwing?

I've more work to do in here, so feel free to raise a github issue about this and assign to me.